Inconsistencies and improvements to SST model

rois1995 opened the pull request:
## Proposed Changes
Hi everyone, 

I have found some inconsistencies with respect to the literature on the implementation of the Menter's SST model. I would like to use this branch as test bench for any corrections/improvements made to the SST model.

Implementation errors found:
- Use of production limiter constant in the clipping of the cross-diffusion term for the computation of the F1 blending function in CTurbSSTVariable.cpp. 
- SST naming conventions (V in V2003m should stay for Vorticity production term.). Not yet assessed.
- Wrong cross-diffusion term included into the residual of w in turb_sources.hpp. It should not be only the positive value as considered in the computation of the F1 blending function. As stated in https://doi.org/10.2514/3.12149.  "CDkw is the
positive portion of the cross-diffusion term in Eq (A2)" pag. 1604. Moreover, a clipping has been introduced for large negative values of this term, as suggested in Peng, Shia-Hui, Peter Eliasson, and Lars Davidson. "Examination of the shear stress transport assumption with a low-Reynolds number k-omega model for aerodynamic flows." Eq 17.
- Boundary conditions errors at inlets, following the Issue #1851. A fix has been proposed following @pcarruscag  suggestions for the supersonic inlet BC.

Changes to SST model proposed:
- Inclusion of non modified versions of SST. In this case I think that more work is needed to be sure that the correct terms are taken into account everywhere in the code.

I've seen the proposed changes to the lower limits of k and w in #2323 and I tried implementing it in my branch. However, if the implementation proposed in the respective PR is preferred then I will change mine. 

## Related Work
#2323 #1851 



## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [X] I am submitting my contribution to the develop branch.
- [ ] My contribution generates no new compiler warnings (try with --warnlevel=3 when using meson).
- [ ] My contribution is commented and consistent with SU2 style (https://su2code.github.io/docs_v7/Style-Guide/).
- [ ] I used the pre-commit hook to prevent dirty commits and used `pre-commit run --all` to format old commits.
- [ ] I have added a test case that demonstrates my contribution, if necessary.
- [ ] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp), if necessary.

bigfooted commented on Common/include/CConfig.hpp:
```diff
@@ -856,7 +856,8 @@ class CConfig {
   nMu_S,                         /*!< \brief Number of species reference S for Sutherland model. */
   nThermal_Conductivity_Constant,/*!< \brief Number of species constant thermal conductivity. */
-  nPrandtl_Lam,                  /*!< \brief Number of species laminar Prandtl number. */
+  nPrandtl_Lam,                  /*!< \brief Number of species 
+  addDoubleOption("FREESTREAM_TURB2LAMVISCRATIO", TurbIntensityAndViscRatioFreeStream[1], 10.0); Prandtl number. */
```
Was this intentionally commented, or...? 
```suggestion
  nPrandtl_Lam,                  /*!< \brief Number of species laminar Prandtl number. */
  addDoubleOption("FREESTREAM_TURB2LAMVISCRATIO", TurbIntensityAndViscRatioFreeStream[1], 10.0); /*!<\brief  Freestream mu_turb to mu_lam viscosity ratio */
```

rois1995 replied:
This is just a wrong copy-paste from me. It should not be there in the first place.

bigfooted commented:
Great contribution, Thanks @rois1995 !

pcarruscag commented:
If you are looking into robustness aspects too you should get in touch with @emaberman and @YairMo, seems like they have some good ideas and between the free time of 3 people a lot more can get done :)

YairMO commented:
Hi,

Regarding the cross-diffusion term (CD) that appears in Omega source term (residual). The SST model (1994/2003) is a high-Reynolds-number  model. Namely, It can not predict correctly the sub-layer region (especially the correct profile of the TKE). Therefore, only a positive contribution is required. Moreover, since the SST model was design as a k-w and k-epsilon blending, the CD term "belongs" only to the k-epsilon "branch", that is why the CD term include the factor "1-F1". However, it may happen, that the factor "1-F1" is not a 100% safe guarantee. It may happen that "1-F1" is not zero in region where the CD term is negative (this happen due numerical errors). To avoid such a situation, it is a good idea to clip the CD term with zero. Otherwise, severe numerical robustness issues may rise. 
Yes, it is different from Menter publications, but I think that clipping  the CD term with zero is completely inline with Menter original idea (that is why, I think,  he was including the factor "1-F1". But again the 1-F1 factor is not 100% percent "safe").

YairMO commented:
Hi,

The use of an Omega production limiter (about the cross-diffusion term) is correct for low-Reynolds-number (LRN) models (the approach described by Peng et al. is very naive; there are other more rigorous treatments). For high-Reynolds-number (HRN) models, the clipping should be zero, keeping the cross-diffusion term positive; thus, the current implementation is correct.

Indeed, it is not exactly as it appears in Menter's original publication. The factor (1-F1) aimed to promise that the cross-diffusion term will be activated only outside the boundary layer, where it is positive (the cross-diffusion term switches its sign deep inside the boundary layer). This was also recognized by Peng et al. (first paragraph above Eq. 17). However, it may happen that the factor (1-F1)=1 where the cross-diffusion term is negative. Usually, it may happen at the wake, very near the airfoil trailing edge, where the upper and lower boundary layers merge. It is due to the imperfection of the F1 function.

To summarize, the current implementation is correct, and it is perfect for HRN models.

YairMO commented:
For the sake of clarity, "current implementation" refers to the current treatment of the production code

emaberman commented:
What YairMO is saying, is that allowing negative cross diffusion values is incorrect for high Reynolds models and should not be an option, this is a fix used for low Reynolds models only

rois1995 commented:
Hi @YairMO, Hi @emaberman ,

thank you very much for your comments. I haven't found any suggestion in literature to clip to only positive values the cross-diffusion term in the w-equation. I understand that it might be more robust, but it is not the standard implementation of the SST model, which is the first thing that we need to achieve. Only then we can build on top of that to improve the robustness of SU2.

Nevertheless, I tried the SWBLI test case and I compared the results across 6 different combinations:

1- develop branch, no changes
2- develop branch, changes to Supersonic_inlet profile as suggested in #1851 
3- my branch, with original CDkw implementation (should give exactly the same result as develop+modified BC)
3- my branch, with original CDkw implementation and using boundary conditions from TMR
4- my branch, with original CDkw implementation and using your suggestions for lower limits for k and w.
5- my branch, allowing negative values of CDkw
6- my branch, allowing negative values of CDkw and using boundary conditions from TMR
7- my branch, allowing negative values of CDkw and using your suggestions for lower limits for k and w.
8- my branch, allowing negative values of CDkw, using boundary conditions from TMR and using your suggestions for lower limits for k and w.

When my branch is used, then the changes to the supersonic inlet BC are already in place.

I haven't achieved convergence with 1, 2 and 3. More precisely, 1 diverged right away (after 30 iterations), while 2 and 3 gave "FGMRES - Orthogonalization Failed" after 900ish iterations.

Here you can see the residuals for the different combinations.

![OrigCDkw](https://github.com/user-attachments/assets/18836f37-5fa5-4523-bee2-8f0c420d4080)

![NegCDkw](https://github.com/user-attachments/assets/4ba446a0-f77d-4f06-bcc2-cdd499839f82)

Unfortunately I will be busy with the AIAA Conference next week, thus I don't know how much I will be able to work on this. The next test case will be the multi-element airfoil.

YairMO commented:
Hi rois1995,

First of all, enjoy your time in Las Vegas. Any paper that you are presenting?

As for our discussion about the cross-diffusion term, I've emailed the "source" (Menter). I believe he will make it clear.
It may be that he will be able to answer only in a while ...

pcarruscag commented:
In my experience, by-the-book implementation of SST is very unstable in real-world meshes and applications.
For example there is a production limit, but it's proportional to k and w, and thus if those start growing for some reason there is nothing preventing the model from exploding.
CD is also tricky due to the 1/w part...

YairMO commented:
I totally agree. On real meshes, Omega usually drops to extremely low values. In cases where the cross-diffusion term is negative (allowed to be negative),  it behaves as a sink term, further amplifying the drop of Omega. A simple addition of this term to the implicit diagonal is insufficient (I tried this). Other more rigorous methods are required (some available in the open literature). 

My main argument is that the factor (1-F1) guarantees only the k-epsilon branch outside the boundary layer, in which the cross-diffusion term is already positive. It may not be so because of numerical errors.

YairMO commented:
Regarding "Fixed Supersonic inlet BC inclusion of TKE," does the energy equation include the contribution of turbulent diffusion?

YairMO commented:
Hi All,

We are lucky, Florian Menter just replied to me.

He agreed that the factor (1-F1) should activate the CD term only at the "k-epsilon" branch, where the CD term is already positive.
Therefore, clipping with zero is in terms of the model and is not incorrect.


Me: A simple boundary layer simulation will reveal that the CD term switch sign is inside the boundary layer and that the F1 function switches from 1 to zero outside the boundary layer. Namely, the CD term should be positive. Florian is unaware of the occasions when it may be negative (even though the model was designed so that this term will be activated when it is positive). 

Best wishes,
Yair

rois1995 commented:
> Regarding "Fixed Supersonic inlet BC inclusion of TKE," does the energy equation include the contribution of turbulent diffusion?

Yeah, I do think so.

rois1995 commented:
I have tested the tripping of the cross-diffusion term also in the residuals on a more complicated test case and indeed it improved convergence. Should I remove this change and use the value of cross-diffusion coming from the computation of F! blending function as was before?

YairMO commented:
Hi,

May I understand "tripping" as "clipping"? 

The cross-diffusion term (that appears in the residual of Omega) is better treated as follows:

2 * (1-F1) * sigma_w2 * rho / Omega * max(grad K * grad Omega, epsilon)

Where "epsilon" could be = 1.e-20 (personally, I'm using epsilon=0)

The term CDkw (that is used to calculate F1) should be implemented in a similar manner:

CDkw = 2 * sigma_w2 * rho / Omega * max(grad K * grad Omega, epsilon)

where epsilon = 1e-20 for the SST1994 and epsilon=1e-10 for the SST2003.

You mentioned that the SU2 implementation of CDkw is not as above?

rois1995 commented:
Yeah, sorry, bad wording from my side. The standard version of SU2 used the value of CDkw as value of the cross-diffusion term in the residual computations. Thus, I think I should revert to that formulation.

Regarding the CDkw, it's implementation in SU2 was made such that the exponent of the epsilon term was equal to minus the production limiter constant of the model. Thus 20 for SST1994 and 10 for SST2003. Is it just a coincidence in the model formulation? In the original paper there is no mention of it being the same variable. This can lead to different results if someone changes the production limiter constant.

pcarruscag commented:
I expect that epsilon to be a simple measure to avoid division by 0, if that lower bound had physical meaning it would have  to be multiplied by some reference factors to make its dimensions appropriate, otherwise SST would not give the same results for the same Reynolds obtained with different rho and mu.

YairMO commented:
You are correct! I have asked Menter about this many, many years back. I don't remember his response, but the formulation given in the TMR is correct (with this fault!)

rois1995 commented:
I have two considerations on the new  boundary conditions: 

1- should the "domain length" in the Omega farfield BC be adimensionalized by the Reynolds length? Otherwise w gets adimensionalized only by the reference velocity.
2- the farfield value of TKE does not take into account for the freestream turbulence intensity. If we want to match the tke computed via $k_{\infty}=w_{\infty} * (\mu_{\infty} * ViscRatio)/\rho_{\infty}$ and the one computed previously as $k_{\infty}=(3/2)*\sqrt((U_{\infty}*TI_{\infty})^2)$, we can either compute the viscosity ratio combining the two equations or just use the one  which explicitly contains the $TI_{\infty}$ disregarding the dependency on $w_{\infty}$. But can we do it? The model as written on the NASA TMR does not explicitly contains the turbulence intensity in the boundary conditions. Maybe it only matters when transition is taken into account.

YairMO commented:
Hi, 

I'm not sure what "new boundary conditions" means. Is someone replacing the present far-field BC of the TKE and Omega? 
I recognize these "new" BCs as the ones given on the TMR (proposed in the original paper).

I find the present setting in SU2 very comfortable and more "intuitive" than that given on the TMR.

rois1995 commented:
So, the two boundary conditions currently implemented in this branch are:

Standard SU2 BCs 
$k_{\infty} = (3.0/2.0) * (U_{\infty} * TI_{\infty})^2$
$w_{\infty} = \rho_{\infty} * k_{\infty} / (\mu_{lam,\infty} * ViscRatio)$
$ViscRatio_{default} = 10.0$
$TI_{\infty, default} = 5.0$

NASA TMR BCs for SST1994 and SST 2003 model
$U_{\infty}/L \leq w_{\infty} \leq 10 U_{\infty}/L$
$10^{-5}U_{\infty}^2/Re_{L} \leq k_{\infty} \leq 10^{-1}U_{\infty}^2/Re_{L}$

where "L is the approximate length of the computational domain," and the combination of the two farfield values should yield a freestream turbulent viscosity between $10^{-5}$ and $10^{-2}$ times freestream laminar viscosity, as cited on the NASA TMR page.

They give very different values of k and w at the farfield. For example, for a the standard test case of the SWBLI with $ViscRatio = 0.01$ and $TI_{\infty} = 5.0$:

Standard SU2 BCs 
$k_{\infty} = 2574.55$ $m^2/s^2$ and $w_{\infty} = 3e10$ $s^{-1}$

NASA TMR BCs
$k_{\infty} = 0.0013$ $m^2/s^2$ and $w_{\infty} = 1563$ $s^{-1}$

I assume that in the end it doesn't matter since the farfield values drop significantly in the standard version of SST.

rois1995 commented:
I am trying to run again the SWBLI V&V case on the SU2 page. However, as it is provided, it does not work with the SU2 standard boundary conditions. In particular, it always diverges due to NaN residuals in the X-Momentum equation. However, when I use the BCs from the NASA TMR page it works just fine (although only the v2003m model). 

Here are the results of the computations performed. 

Three configs are used:
1 - With dimensionless limits $k_{lim} = 10^{-20}$ and $w_{lim} = 10^{-10}$.
2 - Without dimensionless limiters.
3 - With dimensionless limits $k_{lim} = 10^{-30}$ and $w_{lim} = 10^{-20}$.

I am reporting the Skin Friction Coefficient on the lower wall against the Experimental results and the trend of the residual for $\rho$ during the simulations. Five different meshes have been considered.

![ComparisonNewBCsResiduals](https://github.com/user-attachments/assets/d84d7a23-e173-449a-9968-8d8d49b6000f)

![ComparisonNewBCsSFC](https://github.com/user-attachments/assets/243c55c6-e9c7-4335-b889-67244fa81e6b)

The solutions are pretty much identical for the same mesh level, but the cases with the config 3 achieved convergence across all of the meshes. It is interesting how, on finer meshes, the use of dimensionless limits (config 1) actually had worst convergence than the one with std limits (config 2) (up until the finest mesh, where config 2 diverges).

I am going to try to perform the simulations also with the non m variants of the SST model and the full production term. It may require some work since all of the occurrences of the "ComputeStressTensor" have to be changed. Ideally I think that it should be computed only once and  then used that value for all of the other instances. I will focus only on the V2003 model for now as it is the only one that does not diverge.

rois1995 commented:
@YairMO  are these results something that you would expect to obtain when changing the lower limits of k and w? With the new updates to the Reynolds Stress computations I obtain mixed results:

![ComparisonNewBCsResiduals](https://github.com/user-attachments/assets/4c58afcb-881e-4921-ae4b-6ce5b56e903b)

On the finest mesh the dimensionless limits diverge regardless of their value (I also tried decreasing them but it didn't work).

YairMO commented:
Hi, 

To begin with, thank you very much for your efforts and the tremendous work you put into improving the SST model!

It is very difficult for me to judge because I'm using a different time integration than the SU2, which may be critical in this regard. Moreover, my code is non-dimensional, so estimating the influence of the "newBC" and the "lower limits" is difficult for me. Can you tell the values of the "lower limits" relative to the freestream values? Please provide many details about the inputs in your config file (Linear solver, linear solver convergence, CFL, inviscid flux, etc')

YairMO commented:
I found the SWBLI config file. I noticed that it is the NK solver. I would test this case without the NK solver.
Moreover, I've noticed that the entropy fix is very low; how about raising it to 0.2 or using a different flux (JST, AUSMPLUSUP2)? It may sound irrelevant, but it is worth the try.

rois1995 commented:
Changing the entropy fix coefficient (EFC) to 0.2 had the largest effects across the board. It actually degraded the convergence rate, but it removed divergence of the solvers for finer meshes.

![ComparisonNewBCsResiduals_EF0 2](https://github.com/user-attachments/assets/8dbfdaa1-a106-4a54-a364-4d85cefb7eb9)

The converged solutions do not change when changing the EFC, which is desirable.

![ComparisonNewBCsSol_EF0 2](https://github.com/user-attachments/assets/98a01d52-098c-43fb-92d5-39c8285ebccd)

YairMO commented:
Great! This is a non-linear example: Can we say that the convergence pattern is not necessarily directly affected by the turbulence model's lower limits? 
Did you try avoiding the NK solver?

rois1995 commented:
I don't think we can actually say that, or, at least, we can say that it is not the only factor affecting the convergence. The problem is that it is not consistent across the meshes: for sure, using dimensionless limits help, but changing their values has different effects for finer meshes.

![ComparisonNewBCsResiduals_EF0 2_Other](https://github.com/user-attachments/assets/c50ef5b6-3d96-4a24-96ff-3dfdf704f826)

The effect on the residual of the turbulent variables is the same. The most problematic one is always the Omega equation where the residuals remain always large.

![ComparisonNewBCsResiduals_EF0 2_Other_TKE](https://github.com/user-attachments/assets/bdc5383d-37e1-4b1d-a8c2-a2f7edfccb67)

![ComparisonNewBCsResiduals_EF0 2_Other_Omega](https://github.com/user-attachments/assets/3d2b413c-57bc-4692-9b06-73c0065fff56)

The problem is that Omega is too large with respect to all the other variables, especially near the wall (where the residuals remain high).

YairMO commented:
I agree. 
The NK solver is not well suited to solving the turbulence model. Maybe the linear solver alone will be more robust.?

pcarruscag commented:
The NK solver is only applied to the flow equations.

YairMO commented:
Thank you. How about our lower limits (committed):

K_lowLimit = 1E-15 * K_FarField
Omega_lowLimit = 1E-5 * Omega_FarField

rois1995 commented:
It changes somehow the convergence, but again, it doesn't always have the same effect.

![ComparisonNewBCsSol_EF0 2](https://github.com/user-attachments/assets/1ef14d0a-4f98-4cbc-8846-359dbfafe290)

My problem right now is that if I run the v2003m config on the finest mesh with 40 cores then it diverges after 125 iterations. If I run with 20 cores then it does not diverge. The number of cells/core is more then 10k thus I don't think that the domain is overly partitioned.

pcarruscag commented:
If you are using multigrid, it is known to be sensitive to partitioning. You could always use the hybrid parallel mode, that lets you use more cores with less partitioning.

rois1995 commented:
Currently I am not using multigrid. I tried with with 20 cores and 2 threads and it almost results in the same rate of convergence of the simulation with only 20 cores. I also tried going back to a couple of commits, right after the merge of the develop branch and it does not diverge on 40 cores. Starting from that commit I only slightly changed the computation of the stress tensor, which does not involve new mpi communications, thus I don't understand why it should lead to divergence. I will look further into this.

rois1995 commented:
I have tried running 7 consecutive and identical simulations with 4C10T and the residuals are not consistent. Moreover, Run 1 and 5 diverged respectively after 251 and 430 simulations. Here I am showing the residuals of the TKE (I know that NK does not apply to the turbulence solver but it is cleaner to visualize than the residual for the density)

![NonDeterministicRMSk](https://github.com/user-attachments/assets/68208b18-8f81-425c-8672-789576f14c80)

I also tried going back a few commits right before changing the Reynolds stress computation and the situation is even worse.

![NonDeterministicRMSk_PreviousCommit](https://github.com/user-attachments/assets/3f64cd05-dfa4-48cc-9165-b6491481065e)

Here i summarize the convergence of these runs, 1e4 iterations means that it has "converged":

| Run | Iterations | Iterations (old commit)| 
|  :---:  |      :---:       |                  :---:                  |
|    1   |      251       |                1193                 |
|    2   |      1e4       |                1286                 |
|    3   |      1e4       |                  95                   |
|    4   |      1e4       |                 458                  |
|    5   |      430       |                 1e4                  |
|    6   |      1e4       |                 254                  |
|    7   |      1e4       |                 1e4                  |

I might try with the develop branch, but if I recall correctly I was not able to run the test case in the first place. But maybe it was just a random divergence.

rois1995 commented:
I am trying with the .geo file used to create the meshes in the V&V folder. However, I am trying with a finer mesh, with a mesh sizing factor of f = 3.9762. I also tried the develop branch and, even though it always diverges, also there the iterations done are different. I have updated the previous comment.

With the coarser meshes I might have the same problems, but convergence is easier thus I think that I might need more trials. I'll start running some simulation with the 3rd level grid.

pcarruscag commented:
Are you initializing from coarser results? I just checked that the L3 mesh still works fine in develop initialized from L2 results.
The NK solver has some non determinism when using threads especially if the convergence is rough. Using a lot of cores also makes the ILU preconditioner less effective, so you may need to tweak linear solver iterations or CFL to compensate. For reference when I ran the case originally I had a 4 core laptop xD

rois1995 commented:
Every simulation starts from zero. I also tried the third grid level and these are the results for after just four runs

![NonDeterministicRMSk_Level3](https://github.com/user-attachments/assets/7db8e045-1d18-443b-9236-98f85687a22a)

Run 2 diverges while Run 4 converges to the desired residual value (log(RMS(Res_rho)) < -11.5). Running on 4 cores and 10 threads is a problem in this case? Or has the same problems of using 40 cores?

pcarruscag commented:
The ILU issue depends on number of cores regardless of MPI/OpenMP split.
The non-determinism of the NK solver should be worse with more threads.
I recommend initializing from the coarse results.

rois1995 commented:
I tried a couple of things:

- with 20C2T and the results are worse: 4 out of 7 runs diverged, one right away (84 iterations) and the others after more than 1000 iterations. 

- restarting the simulation from the level 3 mesh didn't change the results: 2 out of 7 simulations diverged. 

- running without NK improved a quite a lot, but the inconsistencies remain. Not only that, but something strange can be seen looking at the first hundreds of iterations: the even runs are on top of each other and the same happens for the odd ones.

No Newton-Krylov
![NonDeterministicRMSk_NoNK](https://github.com/user-attachments/assets/5441c7ad-c49c-4a1e-8eec-973e80ee8077)

No Newton-Krylov Zoomed
![NonDeterministicRMSk_NoNKZoom](https://github.com/user-attachments/assets/397d80d1-e270-4424-87a8-43417993f678)

The peak in the residuals that you can see around 3k iterations is different for each run, and leads to slightly different residuals after 10k iterations. 

- Running with 40C1T causes divergence of the simulation at the same iteration for every run (using NK), and in this case the residuals are on top of each other.

I will try having a look into the atomic operations.

pcarruscag commented:
There are two sources for those, `atomicAdd` which is used in dot products with CSysVectors (so it affects the linear solvers). And in the NK solver to compute the norm of the solution.
The other less obvious source is SU2_OMP_CRITICAL, critical sections do not have a guaranteed ordering. We use these to compute the RMS residuals, and those influence the CFL adaptation.
AFAIK the only way around this is to store the result of each thread into a vector visible by all threads, and then have a single thread compute the total.

rois1995 commented:
I can try having a look into this, but I am not sure of how much help I can be. Just from a quick look at the code I can see that the CSYSVEC_PARFOR has a NOWAIT instuction, whereas the SU2_OMP_FOR_STAT doesn't, and these are used right before the atomicAdd call respectively in the CSysVector and in the CNewtonIntegration. But I doubt that's the main cause of large discrepancies when using NK.

pcarruscag commented:
I think NK is just more sensitive because the linear system uses a nastier Jacobian matrix.

rois1995 commented:
Having a better look into the comparison between this branch and the develop one for the simulation with 10C and 1T.

When looking at the solution, the residuals actually improve with this branch everywhere except at the bottom symmetry BC. Don't really know why, I am checking all of the changes to check. 

![ComparisonAbsResRho_MyBranchVsDevelop](https://github.com/user-attachments/assets/9eb98e03-c7a3-453e-98d9-f7769a266fe0)

pcarruscag commented on SU2_CFD/include/numerics_simd/flow/variables.hpp:
```diff
@@ -124,7 +135,8 @@ FORCEINLINE CRoeVariables<nDim> roeAveragedVariables(Double gamma,
   roeAvg.enthalpy = (R*V.j.enthalpy() + V.i.enthalpy()) * D;
-  roeAvg.speedSound = sqrt((gamma-1) * (roeAvg.enthalpy - 0.5*squaredNorm(roeAvg.velocity)));
+  roeAvg.tke = (R*V.j.tke() + V.i.tke()) * D;
+  roeAvg.speedSound = sqrt((gamma-1) * (roeAvg.enthalpy - 0.5*squaredNorm(roeAvg.velocity) - roeAvg.tke));
```
Is enthalpy being computed with tke? If not and then you subtract tke here...
Also please control the scope of this PR, you have another 3 work in progress PRs that took time to review and now are just sitting.

rois1995 replied:
My bad, the enthalpy is computed just as E-P/rho, thus the kinetic energy should not be involved here. Regarding the other WIPs, they all involve the SST model (except for the recent SA one, but that should be finished), thus I think they should wait for this branch first.

rois1995 commented:
As of now the problem is that the supersonic inlet has the correct values of Mach number (5) but on the symmetry BC this does not occur. It seems that on the symmetry BC the problem is the same as it was for the supersonic inlet before, but I still have to find a way to fix this. Moreover, this only happens to the lower symmetry BC, whereas at the top one the Mach is correct.

YairMO commented:
Maybe I'm mixing it up, but I remember there was an issue with the wrong treatment (I'm not sure exactly what it was, perhaps gradient treatment) at Symmetry BC.

rois1995 commented:
I will look into this, but what I found is that the velocity component normal to the bottom symmetry BC is non-zero in my branch.

YairMO commented:
I think this was one of the problems mentioned

rois1995 commented:
Yeah but this does not happen on the results with the develop branch and I have pulled the latest develop commits into this branch, thus it might not be the cause but the consequence of something else.

rois1995 commented:
Part of the problem was that I was considering the tke in the computation of the boundary conditions but not inside the domain since I was using the modified version of the SST model. With this, the vertical velocity at the inlet is in the range of 10^-12. The problem on the bottom symmetry still remains though.

rois1995 commented:
It seems that there is a problem also in the develop branch with the SA model. The Mach number at the lower symmetry BC is higher than the inlet mach number. It does not happen straight away but after some hundreds of iterations.

![LowerSymmetrySA](https://github.com/user-attachments/assets/aaf1ca00-3120-406d-9d6c-00669b9617bd)

YairMO commented:
Hi,

Could you please plot the eddy viscosity? How are the wall distance and tilde{\nu} defined at symmetry BC?

rois1995 commented:
The wall distance is correctly computed. The nu_tilde, however, is higher on the symmetry plane, just like the Mach number. And, again, it happens only at the lower symmetry boundary.

![LowerSymmetrySA_NuTilde](https://github.com/user-attachments/assets/9f2c5a49-9552-47d6-99d1-fe1fad5cb07f)

YairMO commented:
Based on your input (unexpected generation of the eddy viscosity at the lower symmetry BC), I would suspect the wall distance calculation. Can you tell how the wall distance is calculated in the cells above the symmetry section? Consider a cell above the symmetry section. Is its wall distance taken relative to the symmetry part or relative to the wall part?

rois1995 commented:
It is taken relative to the wall part (which you can see as a straight white line. I don't know if the increase in nu_tilde causes an increase of the vertical velocity on the symmetry or it's the opposite.

![LowerSymmetrySA_WD](https://github.com/user-attachments/assets/c6d2878f-e139-4458-8366-d74b4265abd5)

YairMO commented:
Yes, the WD is as it should be; thank you. I also agree: who comes first, the egg or the chicken? I'll try to run this case using my code.

rois1995 commented:
I think the problem is in the fix for the symmetry boundary in PR #2194. I tried with the develop at the last commit, at the commit before the implementation of the correction and at the commit of the implementation. It is clearly visible that something is happening with the symmetry BC correction. I will look further into it

![CommitsComparison_RMSRho](https://github.com/user-attachments/assets/7db41c0e-2885-4e66-bd1c-d01b038430bc)

rois1995 commented:
I tried without the MUSCL reconstruction and with the AC commit and both Mach and nu_tilde profiles are correct at the symmetry plane. @bigfooted  can you comment on this please?

rois1995 commented:
I think I found out the problem. In the correction of the gradients on the symmetry plane, somehow the gradient of the velocity along the normal direction was still not 0. By fixing it then the Mach number and nu_tilde profile are correct. I don't know if I am still missing the correction in the tangential direction, but at least the solution improves.

Here you can see the same graph as before but without the line of the last commit. I have added, instead, a try with the Weighted Least Square (LS) and a first order (FO) (no MUSCL). In the end I have added the line for the correction implemented (Corr Grad). We can see that the behaviour before the implementation of the gradient correction is recovered.

![CommitsComparison_RMSRho_New](https://github.com/user-attachments/assets/dc7c843c-8141-489b-b627-a59cfb2d795d)

pcarruscag commented on SU2_CFD/include/gradients/correctGradientsSymmetry.hpp:
```diff
@@ -61,17 +61,25 @@ inline void correctGradient(const size_t varBegin, const size_t varEnd, const in
+    //   }
+    // }
+
+    // Do I really need all of the previous code? Can't I just merge the two for loops? Maybe I am missing the tangential direction
```
Yes what you are doing neglects that velocity is a vector quantity

rois1995 replied:
The previous implementation was actually correct, I'll revert to that one. Nevertheless something is off.

rois1995 commented:
I've tried reducing the residuals up to -11.5 for the logarithm of the density residual. I've started on the coarsest grid as you saw in the previous plot. I've proceeded on the L2 mesh keeping either the tangential or the first cell spacing constant, using as restart the converged solution on the L1 mesh. The results are essentially the same as with L1 mesh. When using the NK solver straight away the AC branch does not converge, while the BC one easily converged. If I remove the NK solver, both converge. The same has happened on the L3 grid (restart from converged L2 results). 

Here you can see the residuals for the L2 grid (either the case, he outcome is the same)


![CommitsComparison_RMSRho_Grid2](https://github.com/user-attachments/assets/479a0145-dd53-456e-a41d-686037011052)

pcarruscag commented:
Thanks for testing, part of the NK options is a period where instead of NK we use the normal quasi-Newton approach.
I can't give too much detail on this, but we lack a good startup strategy, ideas from papers on NK approached could be useful even for the normal solver.

rois1995 commented:
But what do you think of the changes between the two commits? It doesn't seem that the convergence is improved when using the new symmetry BC implementation.

YairMO commented:
Could you please share your L3 grid (together with your input file)?

pcarruscag commented:
The changes to the symmetry BC were to improve the accuracy, especially when there are walls perpendicular to the symmetry plane.
It's possible the Jacobians can be improved to be more consistent with the RHS and recover some of the convergence rate.

rois1995 commented:
@YairMO in the zip file you can find the .geo file to create the mesh via gmsh, the L3 mesh and the configs file used for simulations with SA.
[MeshAndConfig.zip](https://github.com/user-attachments/files/17174513/MeshAndConfig.zip)

rois1995 commented:
If you see the stagnation in residuals on the L1 meshes while not using the NK solver (line blue and yellow), it is due to a single point in the grid that you can see here. I checked the mesh quality in that point and it doesn't seem to be bad, or at least not worse than the other points.

![StrangeResiduals](https://github.com/user-attachments/assets/78331cc1-bea7-4b21-a12d-d92efda54afd)

YairMO commented:
Hi,

I'm using SU2 8.1.0 but with some differencing in SA implicit treatment (underwork).
I've run the L3 grid with two different input files that differ only by the number of linear solver iterations. Attached is the convergence of the density 

![SWBLI-SU2-Conv](https://github.com/user-attachments/assets/314a1442-b62c-420f-b567-72902c477907)

rois1995 commented:
Impressive! Can you share some info on your implementation? What about the CFL values and the computational time?

YairMO commented:
Thank you. 
The input files that I've used are based on your file. As shown in the figure, mine differs only by the number of linear solver iterations. 
As for the algorithm, I truly apologize; at this moment, I can't share this, but we will soon be able to (we are working on this). Nevertheless, my point in this exercise is to check in your simulation how the number of linear solver iterations affects the convergence of the non-linear problem, especially that of the 77fed34-NK run. 

As for the VFE-2 mesh, I've obtained this mesh from someone else. Please let me check with him first (I believe there will be no problem. Sorry).

rois1995 commented:
Sure, thank you for checking! 

Regarding the SWBLI case, if I understood correctly, you have used your code with NK solver from the start and you changed the number of linear solver iterations, right?

YairMO commented:
Yes, your understanding is correct.
As for the SST ... checking

YairMO commented:
Here are the results obtained using the SST2003 model

![SWBLI-Conv-SST2003](https://github.com/user-attachments/assets/e6f9b1a5-b1d5-4cfd-95b1-023513a712f2)

rois1995 commented:
It seems that yoru implementation affects also the SST model, is it correct? Nevertheless, huge improvement wrt both this branch and develop branch results (density residual stagnates around -10)!

YairMO commented:
Yes, it also affects the SST model. Thanks

YairMO commented:
Hi Rois1995,

I have permission to share the VFE-2 grid; please send me your private email.

rois1995 commented:
That's perfect, thank you! andrea.rausa@polimi.it

rois1995 commented:
I have a few updates. I found out that most of my FGMRES problems in this branch where related to the fact that I was using single precision for the linear system. You can see it in this graphs for the residuals of Rho.

![RMSRho_Mesh_3](https://github.com/user-attachments/assets/185b8195-e415-4dff-9332-cff573c069cc)

The first two curves are in mixed-precision and they stop long before reaching the minimum residual required due to divergence of the linear solver.

Moreover, it seems that the 2003 model here implemented (which, simply put, considers the full reynolds stress tensor for the computation of the production of k and always considers k in the stress tensor and in the thermodynamic variables) has faster convergence and increases the recirculating zone (maybe due to the reduction of turbulence kinetic energy) which is a feature that has also been seen in [DOI:10.1017/aer.2020.93]. 

![SFC_Mesh_3](https://github.com/user-attachments/assets/81a50021-6b58-4a84-9057-a18b5bab023c)

There are some differences between the use of the TMR boundary conditions and the ones used before.

The results improve with respect to the develop in the prediction of the SFC distribution, which has results completely off from the V&V page of SU2

![SFC_Mesh_3_Old](https://github.com/user-attachments/assets/0f4d634b-df31-4f86-a324-d9f67947f1ff)

I may keep on checking if the implementation of the v2003 model is correct or not by searching for other test cases (probably coming from DOI:10.1017/aer.2020.93).

YairMO commented:
How can one switch between single or double precision?

rois1995 commented:
when you execute the meson command use the flag -Denable-mixedprec=true

bigfooted commented:
I am trying to catch up on the developments in this PR but I've lost the overview. Are there still issues that we need to figure out? It looks like convergence is now recovered and results look OK? 
Regarding the symmetry plane, as @pcarruscag mentioned it is improving the accuracy, but not necessarily the convergence. @pcarruscag and I tried to get it to converge as well as before, but no guarantees. Do you still see issues with the new symmetry plane?
Regarding the peak residual that you observed in a single node in the mesh: This is a common observation when using Green-Gauss, which gives O(0) accuracy in mixed mesh nodes where quadrilateral and triangular meshes meet. There are some methods that try to fix this, although none of them are holy grails and the best methods are the most expensive ones.

rois1995 commented:
The 2003 version of SST seems to be working fine. However, the 1994 version is not. I get Mach 7 values on the upper symmetry boundary even on the coarser mesh. I am trying to understand what is happening.

rois1995 commented:
Maybe a stupid question, but I noticed that if I print the coordinates of the i-th point in the CTurbSSTSolver class and the problem is 2D, then the z-coordinate is not 0 but has random values. Does this affects anything in the code or it is not important since the problem is 2D?

pcarruscag commented:
The z coordinate is not allocated for 3D, for you are either reading out of bounds or reading the x or y value from somewhere else.

rois1995 commented:
That's what I thought, thanks for confirming it.

rois1995 commented:
One more thing: should I change the naming scheme for the model version of SST? As it is now, the naming scheme is V+model version but the NASA TMR site says to use the V only for the vorticity form of the production term. Should we remove the V then, or maybe set it as lower case? This will break all of the configs, no problem for me to go through them all and change it though.

YairMO commented:
Chris Rumsey made a tremendous effort to standardize popular RANS turbulence models (via the TMR website).
In my opinion, follow his work. Namely, change the present names to exactly what is given on the TMR.

pcarruscag commented:
No changes that break backward compatibility.

rois1995 commented:
That's what I thought.

bigfooted commented on Common/include/CConfig.hpp:
```diff
@@ -865,7 +865,7 @@ class CConfig {
   nMu_Temperature_Ref,           /*!< \brief Number of species reference temperature for Sutherland model. */
   nMu_S,                         /*!< \brief Number of species reference S for Sutherland model. */
   nThermal_Conductivity_Constant,/*!< \brief Number of species constant thermal conductivity. */
-  nPrandtl_Lam,                  /*!< \brief Number of species laminar Prandtl number. */
```
The original description was correct, for species transport we can have a Prandtl number for each of the species

rois1995 replied:
Yeah I do not recall changing that but for sure it was wrong in my branch. Corrected, thanks!

bigfooted commented on Common/include/CConfig.hpp:
```diff
@@ -865,7 +865,7 @@ class CConfig {
   nMu_S,                         /*!< \brief Number of species reference S for Sutherland model. */
   nThermal_Conductivity_Constant,/*!< \brief Number of species constant thermal conductivity. */
-  nPrandtl_Lam,                  /*!< \brief Number of species laminar Prandtl number. */
+  nPrandtl_Lam,                  /*!< \brief Prandtl number. */
```
```suggestion
  nPrandtl_Lam,                  /*!< \brief Number of species laminar Prandtl number. */
```
